File: pkg/models/workflow.go
package models

import "time"

// Workflow represents a complete automation workflow
type Workflow struct {
	Name        string        `json:"name"`
	Description string        `json:"description"`
	Version     string        `json:"version"`
	Variables   []Variable    `json:"variables,omitempty"`
	Steps       []Step        `json:"steps"`
	OnError     *ErrorHandler `json:"on_error,omitempty"`
}

// Variable defines a workflow-level variable
type Variable struct {
	Name  string      `json:"name"`
	Value interface{} `json:"value"`
	Type  string      `json:"type"` // string, number, boolean, array, object
}

// Step represents a single operation in the workflow
type Step struct {
	ID         string                 `json:"id"`
	Name       string                 `json:"name"`
	Type       string                 `json:"type"` // adb, file, system, web, control
	Enabled    bool                   `json:"enabled"`
	Parallel   bool                   `json:"parallel"`
	Retry      *RetryPolicy           `json:"retry,omitempty"`
	Timeout    int                    `json:"timeout,omitempty"` // in seconds
	Parameters map[string]interface{} `json:"parameters"`
	Conditions []Condition            `json:"conditions,omitempty"`
	Loop       *LoopConfig            `json:"loop,omitempty"`
	OnSuccess  []string               `json:"on_success,omitempty"` // step IDs to run on success
	OnFailure  []string               `json:"on_failure,omitempty"` // step IDs to run on failure
}

// RetryPolicy defines retry behavior for a step
type RetryPolicy struct {
	MaxRetries int  `json:"max_retries"`
	Delay      int  `json:"delay"`   // in seconds
	Backoff    bool `json:"backoff"` // exponential backoff
}

// Condition represents a conditional check
type Condition struct {
	Variable string      `json:"variable"`
	Operator string      `json:"operator"` // equals, not_equals, greater, less, contains, exists
	Value    interface{} `json:"value"`
}

// LoopConfig defines loop behavior for a step
type LoopConfig struct {
	Type     string        `json:"type"` // count, while, for_each
	Count    int           `json:"count,omitempty"`
	Variable string        `json:"variable,omitempty"` // for variable-based loops
	Values   []interface{} `json:"values,omitempty"`   // for for_each loops
}

// ErrorHandler defines error handling behavior
type ErrorHandler struct {
	Action string   `json:"action"`          // continue, stop, retry
	Steps  []string `json:"steps,omitempty"` // step IDs to run on error
	Notify bool     `json:"notify"`
}

// ExecutionContext holds runtime context for workflow execution
type ExecutionContext struct {
	Variables map[string]interface{}
	Results   map[string]interface{}
	StepIndex int
	StartTime time.Time
	Errors    []error
}

File: src/models/workflow.py
"""
Workflow data models for the auto-workflow system.
"""
from dataclasses import dataclass, field
from typing import Dict, List, Any, Optional
from datetime import datetime


@dataclass
class Variable:
    """Represents a workflow-level variable."""
    name: str
    value: Any
    type: str = "string"  # string, number, boolean, array, object


@dataclass
class RetryPolicy:
    """Defines retry behavior for a step."""
    max_retries: int
    delay: int  # in seconds
    backoff: bool = False  # exponential backoff


@dataclass
class Condition:
    """Represents a conditional check."""
    variable: str
    operator: str  # equals, not_equals, greater, less, contains, exists
    value: Any


@dataclass
class LoopConfig:
    """Defines loop behavior for a step."""
    type: str  # count, while, for_each
    count: Optional[int] = None
    variable: Optional[str] = None
    values: Optional[List[Any]] = None


@dataclass
class ErrorHandler:
    """Defines error handling behavior."""
    action: str  # continue, stop, retry
    steps: Optional[List[str]] = None
    notify: bool = False


@dataclass
class Step:
    """Represents a single operation in the workflow."""
    id: str
    name: str
    type: str  # adb, file, system, web, control
    parameters: Dict[str, Any]
    enabled: bool = True
    parallel: bool = False
    retry: Optional[RetryPolicy] = None
    timeout: Optional[int] = None  # in seconds
    conditions: Optional[List[Condition]] = None
    loop: Optional[LoopConfig] = None
    on_success: Optional[List[str]] = None
    on_failure: Optional[List[str]] = None


@dataclass
class ExecutionContext:
    """Holds runtime context for workflow execution."""
    variables: Dict[str, Any] = field(default_factory=dict)
    results: Dict[str, Any] = field(default_factory=dict)
    step_index: int = 0
    start_time: datetime = field(default_factory=datetime.now)
    errors: List[Exception] = field(default_factory=list)


@dataclass
class Workflow:
    """Represents a complete automation workflow."""
    name: str
    description: str
    version: str
    steps: List[Step]
    variables: Optional[List[Variable]] = None
    on_error: Optional[ErrorHandler] = None
